Source/WebCore/rendering/RenderElement.cpp

```diff
@@ -900,18 +900,15 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
         }
 
         // Keep layer hierarchy visibility bits up to date if visibility or skipped content state changes.
-        bool wasVisible = m_style.usedVisibility() == Visibility::Visible && !m_style.hasSkippedContent();
-        bool willBeVisible = newStyle.usedVisibility() == Visibility::Visible && !newStyle.hasSkippedContent();
-        if (wasVisible != willBeVisible) {
+        if ((m_style.usedVisibility() == Visibility::Visible) != (newStyle.usedVisibility() == Visibility::Visible)) {
+            if (CheckedPtr layer = enclosingLayer())
+                layer->dirtyVisibleContentStatus();
```

Comment:
we don't worry about over invalidation here? The content may still have the same perceived visibility when going from `visibility: visible` to `visibility: hidden` when inside skipped content?

Reply:
This is only dirtying the 'is visible' flag, so that we recompute visibility status next time we update descendant dependant flags. That may very well end up not being a change, and not cause any further invalidation of expensive things.

That said, unrelated to this code, changing `visibility` causes a `StyleDifference::Repaint` hint, so the problem still exists. I guess all the various types of visibility should be combined into a `StyleDifferenceContextSensitiveProperty` so that we can compute the final state and only invalidate as necessary.